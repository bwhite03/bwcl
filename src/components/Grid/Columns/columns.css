.mikto-table-modal {
  background-color: #afaeae;
  color: white;
}

.mikto-table-modal-title {
  border-bottom: 3px solid black;
  padding: 3px;
}

.mikto-table-modal-body {
  padding: 8px;
}

.mikto-table-modal-line {
  min-height: 25px;
}

.mikto-table-modal-line-title {
  margin-left: 8px;
}
