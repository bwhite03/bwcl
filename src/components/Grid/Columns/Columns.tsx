import React, { useState } from "react";
import { createPortal } from "react-dom";
import { TableHeader } from "../DataGrid";
import "./columns.css";

export interface ColumnProps<T> {
  open: boolean;
  divId: string;
  style?: React.CSSProperties;
  headers: string[];
  checkedColumns: string[];
  handleCheckClick: (e: string, checked: boolean) => void;

  identifier: string;
  data: T[];
}

function Columns<T>(props: ColumnProps<T>) {
  const { open, divId, style, headers, checkedColumns, handleCheckClick } =
    props;

  const div = document.getElementById(divId);

  const amIChecked = (title: keyof T): boolean => {
    return checkedColumns.includes(title as string);
  };

  if (open && div) {
    return createPortal(
      <div className="mikto-table-modal" style={style}>
        <div className="mikto-table-modal-title">Columns</div>
        <div className="mikto-table-modal-body">
          {headers.map((item: any, idx: number) => {
            if (item.visible && item.visible !== false) {
              if (!checkedColumns.includes(item.title)) {
                checkedColumns.push(item.title);
              }
            } else if (item.visible === undefined) {
              if (!checkedColumns.includes(item.title)) {
                checkedColumns.push(item.title);
              }
            }
            return (
              <div className="mikto-table-modal-line" key={`mc-c-${idx}`}>
                <input
                  className="mikto-table-modal-check"
                  type="checkbox"
                  checked={amIChecked(item.title)}
                  onChange={(e) =>
                    handleCheckClick(item.title, e.target.checked)
                  }
                />
                <span className="mikto-table-modal-line-title">
                  {item.title}
                </span>
              </div>
            );
          })}
        </div>
      </div>,
      div
    );
  } else {
    return null;
  }
}

export default Columns;
