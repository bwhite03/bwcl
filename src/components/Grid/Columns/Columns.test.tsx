import React from "react";
import { render, screen } from "@testing-library/react";
import Columns from "./Columns";
import userEvent from "@testing-library/user-event";

const testData = [
  {
    id: 1,
    storeName: "IGA 001",
    storeNumber: "001",
    termCount: 3,
  },
  {
    id: 2,
    storeName: "IGA 002",
    storeNumber: "002",
    termCount: 4,
  },
  {
    id: 3,
    storeName: "IGA 003",
    storeNumber: "003",
    termCount: 6,
  },
];

const headers = [
  {
    columnName: "id",
    title: "ID",
    visible: true,
    style: {
      textAlign: "center",
    },
  },
  {
    columnName: "storeName",
    title: "Name",
    style: {
      textAlign: "left",
    },
  },
  {
    columnName: "storeNumber",
    title: "#",
    style: {
      textAlign: "center",
    },
  },
  {
    columnName: "termCount",
    title: "Terms",
    style: {
      textAlign: "center",
    },
    visible: false,
  },
];

const header = {
  columnName: "id",
  title: "ID",
  visible: true,
  style: {
    textAlign: "center",
  },
};

describe("Columns", () => {
  test("should render correctly", () => {
    const testFn = jest.fn;
    document.body.innerHTML = '<div id="col-1"></div>';
    const frag = render(
      <Columns
        open={true}
        divId="col-1"
        headers={headers as unknown as string[]}
        checkedColumns={[]}
        handleCheckClick={testFn}
        // @ts-ignore
        header={header}
        identifier={"testid"}
        data={testData}
      />
    );

    expect(frag).toMatchSnapshot();
  });
  test("should allow a header to be clicked", () => {
    const testFn = jest.fn;
    document.body.innerHTML = '<div id="col-1"></div>';
    const frag = render(
      <Columns
        open={true}
        divId="col-1"
        headers={headers as unknown as string[]}
        checkedColumns={[]}
        handleCheckClick={testFn}
        // @ts-ignore
        header={header}
        identifier={"testid"}
        data={testData}
      />
    );

    const check: HTMLInputElement = screen.getByTestId(
      "mikto-table-check-testid-1"
    );
    userEvent.click(check);
    expect(check.checked).toBeTruthy();
  });
  test("should test if a column is already checked", () => {
    const testFn = jest.fn;
    document.body.innerHTML = '<div id="col-1"></div>';
    const frag = render(
      <Columns
        open={true}
        divId="col-1"
        headers={headers as unknown as string[]}
        checkedColumns={["ID"]}
        handleCheckClick={testFn}
        // @ts-ignore
        header={header}
        identifier={"testid"}
        data={testData}
      />
    );
  });
  test("should test if a column is already checked but does not have visible as an attribute", () => {
    const testFn = jest.fn;
    document.body.innerHTML = '<div id="col-1"></div>';
    const frag = render(
      <Columns
        open={true}
        divId="col-1"
        headers={headers as unknown as string[]}
        checkedColumns={["Name"]}
        handleCheckClick={testFn}
        // @ts-ignore
        header={header}
        identifier={"testid"}
        data={testData}
      />
    );
  });
});
