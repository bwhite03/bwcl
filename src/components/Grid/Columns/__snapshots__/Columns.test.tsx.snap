// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Columns should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="col-1"
    >
      <div
        class="mikto-table-modal"
      >
        <div
          class="mikto-table-modal-title"
        >
          Columns
        </div>
        <div
          class="mikto-table-modal-body"
        >
          <div
            class="mikto-table-modal-line"
          >
            <input
              checked=""
              class="mikto-table-modal-check"
              data-testid="mikto-table-check-testid-0"
              type="checkbox"
            />
            <span
              class="mikto-table-modal-line-title"
            >
              ID
            </span>
          </div>
          <div
            class="mikto-table-modal-line"
          >
            <input
              checked=""
              class="mikto-table-modal-check"
              data-testid="mikto-table-check-testid-1"
              type="checkbox"
            />
            <span
              class="mikto-table-modal-line-title"
            >
              Name
            </span>
          </div>
          <div
            class="mikto-table-modal-line"
          >
            <input
              checked=""
              class="mikto-table-modal-check"
              data-testid="mikto-table-check-testid-2"
              type="checkbox"
            />
            <span
              class="mikto-table-modal-line-title"
            >
              #
            </span>
          </div>
          <div
            class="mikto-table-modal-line"
          >
            <input
              class="mikto-table-modal-check"
              data-testid="mikto-table-check-testid-3"
              type="checkbox"
            />
            <span
              class="mikto-table-modal-line-title"
            >
              Terms
            </span>
          </div>
        </div>
      </div>
    </div>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
